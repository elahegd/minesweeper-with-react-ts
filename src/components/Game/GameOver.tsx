import { FC, memo } from 'react';
import styled from '@emotion/styled';

export interface GameOverProps extends FrameProps {
    onClick: () => void;
}

export const GameOver:FC<GameOverProps> = memo(({ onClick, isWin }) => {
  return (
    <Frame onClick={ onClick } isWin={isWin}>
        {isWin ? '😎' : '🙁'}
    </Frame>
  )
});

// Stryker disable next-line StringLiteral
GameOver.displayName = "GameOver";

interface FrameProps {
    isWin: boolean;
}

const Frame = styled.div<FrameProps>`
    top: 50%;
    left: 50%;
    z-index: 11;
    width: 8vw;
    height: 8vw;
    font-size: 4vw;
    text-align: center;
    user-select: none;
    cursor: pointer;
    line-height: 8vw;
    position: absolute;
    border-radius: 50%;
    pointer-events: auto;
    background-color: #d1d1d1;
    transform: translate(-50%, -50%);
    border: 1px solid rgba(51, 51, 51, 0.25);
`;
