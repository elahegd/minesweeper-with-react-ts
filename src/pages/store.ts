import { configureStore } from "@reduxjs/toolkit";
import { reducer } from "../module/GameWithRedux";

export const store = configureStore({
    reducer: {
        game: reducer
    }
});

export type RootState = ReturnType<typeof store.getState>;