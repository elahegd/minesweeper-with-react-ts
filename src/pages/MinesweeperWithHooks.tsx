import React, { FC } from 'react';

import { useParams } from "react-router-dom";

import { GameLayout } from '../components/Game';
import  Top  from "../components/Top";
import { GameWithHooks } from '../module/GameWithHooks';

const MinesweeperWithHooks: FC = () => {
    const { username } = useParams<{ username?: string }>();

    return (
        <GameLayout 
            top={
                <Top feature="Flag" firstAction="ctrl" secondAction="click">
                    Minesweeper with React Hooks special for you: {username && ` ${username}`}
                </Top>
            }
        >
            <GameWithHooks />
        </GameLayout>
    );
}

export default MinesweeperWithHooks;