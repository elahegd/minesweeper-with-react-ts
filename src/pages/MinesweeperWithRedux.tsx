import React, { FC } from 'react';

import { GameLayout } from '../components/Game';
import  Top  from "../components/Top";
import { GameWithRedux } from '../module/GameWithRedux';

const MinesweeperWithRedux: FC = () => {
    return (
        <GameLayout 
            top={
                <Top feature="Flag" firstAction="ctrl" secondAction="click">
                    Minesweeper with React and Redux uses useReducer
                </Top>
            }
        >
            <GameWithRedux />
        </GameLayout>
    );
}

export default MinesweeperWithRedux;