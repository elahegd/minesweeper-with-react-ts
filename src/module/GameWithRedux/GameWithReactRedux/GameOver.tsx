import React, { useCallback, FC } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { RootState } from "../../../pages/store";
import { actions } from "../index";

import { GameOver as GameOverComponent } from "../../../components/Game/GameOver";

export const GameOver: FC = () => {
    const dispatch = useDispatch();

    const { isWin, isGameOver } = useSelector(({ game: { isWin, isGameOver } }: RootState) => ({
        isWin, isGameOver
    }));

    const onReset = useCallback(
        () => dispatch(actions.reset()), 
        []
    )

    return (
        <>
            {isGameOver && <GameOverComponent onClick={onReset} isWin={isWin} />}
        </>
    )
}
