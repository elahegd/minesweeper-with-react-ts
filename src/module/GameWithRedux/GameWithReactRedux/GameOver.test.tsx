import React from 'react';
import { Provider } from 'react-redux';
import { render, screen } from "@testing-library/react";
import userEvent from '@testing-library/user-event';
import { useDispatch, useSelector } from 'react-redux';

import { store } from '../../../pages/store';

import { GameOver } from './GameOver';

jest.mock("react-redux", () => ({
    __esModule: true,
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn(),
    useDispatch: jest.fn()
}))

describe("GameOver test cases", () => {
    it("GameOver renders correctly", () => {
        const mockDispatch = jest.fn();

        (useDispatch as jest.Mock).mockReturnValue(mockDispatch);
        (useSelector as jest.Mock).mockReturnValue({
            isGameOver: true,
            isWin: false,
        })
        const { asFragment } = render(
            <Provider store={store}>
                <GameOver />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();

        userEvent.click(screen.getByText('🙁'));
        expect(mockDispatch).toHaveBeenCalled();
    });

    it("GameOver with win check", () => {
        const mockDispatch = jest.fn();

        (useDispatch as jest.Mock).mockReturnValue(mockDispatch);
        (useSelector as jest.Mock).mockReturnValue({
            isGameOver: true,
            isWin: true,
        })
        const { asFragment } = render(
            <Provider store={store}>
                <GameOver />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();

        userEvent.click(screen.getByText('😎'));
        expect(mockDispatch).toHaveBeenCalled();
    });
})