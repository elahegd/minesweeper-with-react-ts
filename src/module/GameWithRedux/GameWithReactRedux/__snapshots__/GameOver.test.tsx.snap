// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameOver test cases GameOver renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-ap4m5y"
  >
    🙁
  </div>
</DocumentFragment>
`;

exports[`GameOver test cases GameOver with win check 1`] = `
<DocumentFragment>
  <div
    class="css-ap4m5y"
  >
    😎
  </div>
</DocumentFragment>
`;
